### PR TITLE
feat(servicediscovery): added privatednsnamespace API and controller …

### PR DESCRIPTION
**Pull request by blut**
…for AWS servicediscovery

Signed-off-by: Hannes Blut <hannes.blut-extern@deutschebahn.com>
(external expert on behalf of DB Netz AG)

### Description of your changes

* Added generated API code and controllers for AWS servicediscovery
* Added resource type: PrivateDNSNamespace

Fixes # https://github.com/crossplane/provider-aws/issues/562

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Testing has thus far been done manually to establish a first servicediscovery controller and functional requirements.
Unit tests will follow after the resource handling has been discussed

### CODE REVIEW

File: apis/servicediscovery/v1alpha1/custom_types.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package v1alpha1
+
+// CustomServiceParameters are custom parameters for Services.
+type CustomServiceParameters struct{}
+
+// CustomPrivateDNSNamespaceParameters are custom parameters for PrivateDNSNamespaces.
+type CustomPrivateDNSNamespaceParameters struct {
+	// HostedZoneID *string
+	// HTTPName     *string
```
Comment (MisterMX) [inline apis/servicediscovery/v1alpha1/custom_types.go:9]: These commented lines should be removed

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	cpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/servicediscovery/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+func (e *external) observeImpl(ctx context.Context, cpm cpresource.Managed) (managed.ExternalObservation, error) { // nolint:gocyclo
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:19]: You should use a seperate `hook` type where you define your custom functions and pass them to `external` which is auto-generated and should not be modified.

Then you can rename this to `observe` as well.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	ret := managed.ExternalObservation{}
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:20]: I think it would increase readability if a `managed.ExternalObservation` is created for every return statement. It is a bit hard to follow the state of the variable.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	cr, ok := cpm.(*svcapitypes.PrivateDNSNamespace)
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:21]: I think `mg` is the convention for `cpresource.Managed` variables (at least that's how it is named in other controllers. So I would suggest to stick to it.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	if cr.Status.AtProvider.OperationID == nil || *cr.Status.AtProvider.OperationID == "" {
+		return ret, nil
+	}
+	opInput := &svcsdk.GetOperationInput{
+		OperationId: cr.Status.AtProvider.OperationID,
+	}
+
+	opReqResp, err := e.client.GetOperationWithContext(ctx, opInput)
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:37]: I think it should be `opReq` and `opResp`, shouldn't it?

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		return ret, errors.Wrap(err, "get-operation failed")
+	}
+
+	if *opReqResp.Operation.Status == "PENDING" || *opReqResp.Operation.Status == "SUBMITTED" {
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:42]: `opReqResp.Operation.Status` is a pointer and might be nil in some edge cases. This will cause a panic if you try to deref it.
Better to use `awsclient.StringValue`

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+	nsInput := &svcsdk.GetNamespaceInput{
+		Id: namespaceID,
+	}
+	nsReqResp, err := e.client.GetNamespaceWithContext(ctx, nsInput)
+	if err != nil {
+		if ActualIsNotFound(err) || IsDuplicateRequest(err) {
+			cr.Status.SetConditions(xpv1.Unavailable())
+		}
+		return ret, awsclients.Wrap(cpresource.IgnoreAny(err, ActualIsNotFound, IsDuplicateRequest), "get-namespace failed")
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:69]: Error messages should be constants, not magic strings

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+		ret.ResourceLateInitialized = true
+	}
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	lateInitialize(&cr.Spec.ForProvider, nsReqResp)
+	ret.ResourceLateInitialized = ret.ResourceLateInitialized || !cmp.Equal(&cr.Spec.ForProvider, currentSpec)
+
+	upToDate, err := e.isUpToDate(cr, nsReqResp)
+	if err != nil {
+		return ret, errors.Wrap(err, "isUpToDate check failed")
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:85]: Error messages should be constants, not magic strings

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	return ret, nil
+}
+
+func (e *external) updateImpl(ctx context.Context, mg cpresource.Managed) (managed.ExternalUpdate, error) {
+	// Update always returns an error, since the servicediscovery API does not allow updating namespaces
+	return managed.ExternalUpdate{}, nil
+}
+
+func (e *external) isUpToDate(cr *svcapitypes.PrivateDNSNamespace, resp *svcsdk.GetNamespaceOutput) (bool, error) { // nolint:unparam
+	if meta.GetExternalName(cr) != *resp.Namespace.Id {
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:98]: Use `awsclient.StringValue` instead of raw deref

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	// one could use the HostedZone information to verify the VPC configuration
+	// through Route53 HostedZone records
+
+	// tags := map[string]string
+	// if cmp.Equal(cr.Spec.ForProvider.Tags, tags) {
+	// Where does the servicediscovery API provide Tag information?
+
+	// if cr.Spec.ForProvider.Region
+	// Region information is not available through servicediscovery API
+	return true, nil
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:120]: Error in `isUpToDate` is always nil, so it can be removed.

File: pkg/controller/servicediscovery/privatednsnamespace/setup.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/servicediscovery/v1alpha1"
+)
+
+// SetupPrivateDNSNamespace adds a controller that reconciles PrivateDNSNamespaces.
+func SetupPrivateDNSNamespace(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
+	name := managed.ControllerName(svcapitypes.PrivateDNSNamespaceGroupKind)
+	opts := []option{
+		func(e *external) {
+			e.observe = e.observeImpl
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/setup.go:24]: It would be best to create a `hook` type that has `observe`, `create`, `update` and `delete`.

See https://github.com/crossplane/provider-aws/blob/master/pkg/controller/secretsmanager/secret/setup.go as example

File: pkg/controller/servicediscovery/privatednsnamespace/setup.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+		For(&svcapitypes.PrivateDNSNamespace{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.PrivateDNSNamespaceGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+// ActualIsNotFound reimplements IsNotFound which doesn't do it's job
+func ActualIsNotFound(err error) bool {
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/setup.go:45]: Those two functions don't belong to `setup.go` and should rather be in `hooks.go` where they are used.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+		return ret, errors.Wrap(err, "isUpToDate check failed")
+	}
+
+	ret.ResourceUpToDate = upToDate
+	return ret, nil
+}
+
+func (e *external) updateImpl(ctx context.Context, mg cpresource.Managed) (managed.ExternalUpdate, error) {
+	// Update always returns an error, since the servicediscovery API does not allow updating namespaces
+	return managed.ExternalUpdate{}, nil
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:94]: In contrast to the comment `updateImpl` does not return an error

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+func lateInitialize(forProvider *svcapitypes.PrivateDNSNamespaceParameters, nsReqResp *svcsdk.GetNamespaceOutput) {
+	if nsReqResp == nil {
+		return
+	}
+
+	if forProvider.Description == nil && nsReqResp.Namespace.Description != nil {
+		forProvider.Description = nsReqResp.Namespace.Description
+	}
+
+	if forProvider.CreatorRequestID == nil && nsReqResp.Namespace.CreatorRequestId != nil {
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:149]: I think you can remove the second nil check from both if's. `Description` and `CreatorRequestID` will be set to nil anyway.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+	cr, ok := cpm.(*svcapitypes.PrivateDNSNamespace)
+	if !ok {
+		return ret, errors.New(errUnexpectedObject)
+	}
+	externalName := meta.GetExternalName(cr)
+	if externalName == "" {
+		return ret, nil
+	}
+
+	if cr.Status.AtProvider.OperationID == nil || *cr.Status.AtProvider.OperationID == "" {
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:30]: Might wanna use `awsclient.StringValue(cr.Status.AtProvider.OperationID) == ""`

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		return ret, errors.Wrap(err, "get-operation failed")
+	}
+
+	if *opReqResp.Operation.Status == "PENDING" || *opReqResp.Operation.Status == "SUBMITTED" {
+		ret.ResourceExists = true
+		return ret, nil
+	}
+
+	if *opReqResp.Operation.Status == "FAIL" {
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:47]: Better use `awsclient.StringValue`

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+		if cr.GetCondition(xpv1.TypeReady).Reason != xpv1.ReasonDeleting {
+			ret.ResourceExists = true
+		}
+		cr.Status.SetConditions(xpv1.Unavailable().WithMessage(*opReqResp.Operation.ErrorMessage))
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:51]: Check for nil before deref

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		if ActualIsNotFound(err) || IsDuplicateRequest(err) {
+			cr.Status.SetConditions(xpv1.Unavailable())
+		}
+		return ret, awsclients.Wrap(cpresource.IgnoreAny(err, ActualIsNotFound, IsDuplicateRequest), "get-namespace failed")
+	}
+	ret.ResourceExists = true
+
+	cr.Status.SetConditions(xpv1.Available())
+	if externalName != *namespaceID {
```
Comment (MisterMX) [inline pkg/controller/servicediscovery/privatednsnamespace/hooks.go:74]: Better use `awsclient.StringValue`

File: apis/servicediscovery/v1alpha1/generator-config.yaml [hunk 19]
```diff
@@ -0,0 +1,6 @@
+ignore:
+  resource_names:
+    - PublicDNSNamespace
+    - Service
+    - HTTPNamespace
+# How to add +immutable to required fields Region, Name, and VPC
```
Comment (MisterMX) [inline apis/servicediscovery/v1alpha1/generator-config.yaml:6]: Comment should be removed

**Review by MisterMX — CHANGES_REQUESTED** [review]

File: apis/servicediscovery/v1alpha1/custom_types.go (hunk 1 above)
Comment (blut) [reply apis/servicediscovery/v1alpha1/custom_types.go]: Removed commented parameters

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 2 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: updated code to have separate hook struct

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 3 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: I'm not convinced that it helps readability, but I've changed the code to create the external observation on each return

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 4 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: done

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 5 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: done

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 6 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: moved all string pointer dereferences into awsclient.StringValue calls

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 7 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: This is also practice in some of the other controllers I used for reference. I've changed it to a constant.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 8 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: Error does not exist anymore

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 9 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: done

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 10 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: done

File: pkg/controller/servicediscovery/privatednsnamespace/setup.go (hunk 11 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/setup.go]: Thanks for the reference! I've updated the code to use a hooks struct.

File: pkg/controller/servicediscovery/privatednsnamespace/setup.go (hunk 12 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/setup.go]: As far as I'm concerned they should be part of the generated code, which is why I've placed them in the setup.go.
Moved to hooks.go.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 13 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: This function was initially intended to always return an error but has since been reduced to this state. The function has been removed, as it double the nopUpdate function.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 14 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: Again, referenced from another controller, but I agree. Updated without the second test.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 15 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: done

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 16 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: opStatus is now determined apriori with StringValue.

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 17 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: done

File: pkg/controller/servicediscovery/privatednsnamespace/hooks.go (hunk 18 above)
Comment (blut) [reply pkg/controller/servicediscovery/privatednsnamespace/hooks.go]: done

File: apis/servicediscovery/v1alpha1/generator-config.yaml (hunk 19 above)
Comment (blut) [reply apis/servicediscovery/v1alpha1/generator-config.yaml]: How would someone updating the code know that I'd previously added the immutable keyword to required fields in the generated code?

**Comment by blut** [comment]
Squashed all commits

**Review by muvaf — APPROVED** [review]
Apologies for late review. There are a couple of things that would be good to fix but I'll merge this PR and open another one to fix them since this PR has been up for a long time now.
